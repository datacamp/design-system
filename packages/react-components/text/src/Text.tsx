import './sideEffects/injectRegularFontFace';

import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { css } from '@emotion/core';
import { mapKeys } from 'lodash';
import React from 'react';

interface TextProps {
  children: string;
  className?: string;
  dataAttributes?: { [key: string]: string };
}

const baseTextStyle = css({
  color: tokens.color.opaque.grey.value.rgb,
  fontFamily: [
    tokens.asset.font.sansSerif.attributes.fallback,
    tokens.asset.font.sansSerif.value,
  ],
  fontSize: tokens.size.font.base.value,
  fontStyle: 'normal',
  fontWeight: tokens.size.fontWeight.regular.value,
  lineHeight: tokens.size.font.base.value,
});

export const Text: React.FC<TextProps> = ({
  children,
  className,
  dataAttributes,
}: TextProps) => {
  const parsedDataAttributes = mapKeys(
    dataAttributes,
    (_, key) => `data-${key}`
  );

  return (
    <span css={baseTextStyle} className={className} {...parsedDataAttributes}>
      {children}
    </span>
  );
};

export default Text;
