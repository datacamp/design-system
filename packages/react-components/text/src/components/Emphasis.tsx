/* eslint-disable global-require */
import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { childrenOfType, computeDataAttributes } from '@datacamp/waffles-utils';
import { css } from '@emotion/react';
import PropTypes from 'prop-types';
import React, { ReactElement, ReactNode } from 'react';

import baseStyle from '../baseStyle';

interface EmphasisProps {
  /**
   * The content to display. Can contain a combination of strings, Text
   * components, and Strong Components.
   */
  children: ReactNode;
  /**
   * Sets the css class of the rendered element. Can be used to apply custom
   * styles.
   */
  className?: string;
  /**
   * As with all the other waffles components, dataAttributes can be used to set
   * data- html attributes on the element.
   */
  dataAttributes?: { [key: string]: string };
}

const emphasisStyle = css(baseStyle, {
  fontStyle: 'italic',
  fontWeight: tokens.fontWeight.regular.value,
  lineHeight: tokens.size.font[300].value,
});

const Emphasis = ({
  children,
  className,
  dataAttributes,
}: EmphasisProps): ReactElement => {
  const parsedDataAttributes = computeDataAttributes(dataAttributes);

  return (
    <em className={className} css={emphasisStyle} {...parsedDataAttributes}>
      {children}
    </em>
  );
};

export default Emphasis;

const validChildType = PropTypes.oneOfType([
  childrenOfType(require('./Text')),
  childrenOfType(require('./Strong')),
  childrenOfType(require('../alternateComponents/PlainString')),
  PropTypes.string,
  PropTypes.number,
]);

Emphasis.propTypes = {
  children: PropTypes.oneOfType([
    validChildType,
    PropTypes.arrayOf(validChildType),
  ]),
};
