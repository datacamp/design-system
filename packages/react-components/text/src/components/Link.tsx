import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { computeDataAttributes } from '@datacamp/waffles-utils';
import { css } from '@emotion/core';
import React, { ReactElement } from 'react';

import baseStyle from '../baseStyle';

interface LinkProps {
  /**
   * The content of the Link.
   */
  children: string;
  /**
   * Sets the css class of the rendered element. Can be used to apply custom
   * styles.
   */
  className?: string;
  /**
   * As with all the other waffles components, dataAttributes can be used to set
   * data- html attributes on the element.
   */
  dataAttributes?: { [key: string]: string };
  /**
   * The url to navigate to, when the user clicks.
   */
  href: string;
}

const linkStyle = css(baseStyle, {
  color: tokens.color.opaque.primary.value.hex,
  cursor: 'pointer',
  fontSize: tokens.size.font[300].value,
  fontWeight: tokens.fontWeight.regular.value,
  lineHeight: tokens.size.font.base.value,
  textDecoration: 'underline',
});

const InternalLink = ({
  className,
  dataAttributes,
  children,
  innerRef,
  href,
}: LinkProps & { innerRef?: React.Ref<HTMLAnchorElement> }): ReactElement => {
  const parsedDataAttributes = computeDataAttributes(dataAttributes);
  return (
    <a
      className={className}
      css={linkStyle}
      href={href}
      ref={innerRef}
      {...parsedDataAttributes}
    >
      {children}
    </a>
  );
};

const Link = React.forwardRef<HTMLAnchorElement, LinkProps>((props, ref) => (
  <InternalLink {...props} innerRef={ref} />
));

export default Link;
