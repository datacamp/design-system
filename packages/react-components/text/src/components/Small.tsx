import tokens from '@datacamp/waffles-tokens';
import { computeDataAttributes } from '@datacamp/waffles-utils';
import { css } from '@emotion/react';
import React from 'react';

import baseStyle from '../baseStyle';

interface SmallProps {
  /**
   * The content to display. Can contain a combination of strings, Strong
   * components, and Emphasis Components.
   */
  children: React.ReactNode;
  /**
   * Sets the css class of the rendered element. Can be used to apply custom
   * styles.
   */
  className?: string;
  /**
   * As with all the other waffles components, dataAttributes can be used to set
   * data- html attributes on the element.
   */
  dataAttributes?: { [key: string]: string };
}

const smallStyle = css(baseStyle, {
  fontSize: tokens.fontSizes.small,
  lineHeight: tokens.lineHeights.relaxed,
});

const Small = ({
  children,
  className,
  dataAttributes,
}: SmallProps): JSX.Element => {
  const parsedDataAttributes = computeDataAttributes(dataAttributes);

  return (
    <small className={className} css={smallStyle} {...parsedDataAttributes}>
      {children}
    </small>
  );
};

export default Small;
