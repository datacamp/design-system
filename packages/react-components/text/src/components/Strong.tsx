import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { css } from '@emotion/core';
import React from 'react';

import baseStyle from '../baseStyle';
import computeDataAttributes from '../computeDataAttributes';

interface StrongProps {
  children: string;
  className?: string;
  dataAttributes?: { [key: string]: string };
}

const strongStyle = css(baseStyle, {
  fontWeight: tokens.size.fontWeight.bold.value,
  lineHeight: tokens.size.font.base.value,
});

const Strong: React.FC<StrongProps> = ({
  children,
  className,
  dataAttributes,
}) => {
  const parsedDataAttributes = computeDataAttributes(dataAttributes);

  return (
    <strong className={className} css={strongStyle} {...parsedDataAttributes}>
      {children}
    </strong>
  );
};

export default Strong;
