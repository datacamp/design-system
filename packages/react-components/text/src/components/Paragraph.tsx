import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { css } from '@emotion/core';
import React from 'react';

import baseStyle from '../baseStyle';
import computeDataAttributes from '../computeDataAttributes';
import ssrSafeFirstChildSelector from '../ssrSafeFirstChildSelector';
import validateChildrenProp from '../validateChildrenProp';
import Emphasis from './Emphasis';
import Small from './Small';
import Strong from './Strong';
import Text from './Text';

interface ParagraphProps {
  children: string | (string | React.ReactElement)[];
  className?: string;
  dataAttributes?: { [key: string]: string };
}

const paragraphStyle = css(baseStyle, {
  fontSize: tokens.size.font[300].value,
  fontWeight: tokens.size.fontWeight.regular.value,
  lineHeight: tokens.size.lineHeight.base.value,
  margin: 0,
  marginTop: tokens.size.spatial.size[8].value,
  [ssrSafeFirstChildSelector]: {
    marginTop: 0,
  },
});

const Paragraph: React.FC<ParagraphProps> = props => {
  validateChildrenProp(props, 'Paragraph', [
    Strong,
    Text,
    Small,
    Emphasis,
    'br',
  ]);
  const { children, className, dataAttributes } = props;
  const parsedDataAttributes = computeDataAttributes(dataAttributes);

  return (
    <p className={className} css={paragraphStyle} {...parsedDataAttributes}>
      {children}
    </p>
  );
};

export default Paragraph;
