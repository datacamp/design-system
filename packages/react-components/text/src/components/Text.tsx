import tokens from '@datacamp/waffles-tokens/lib/future-tokens.json';
import { css } from '@emotion/core';
import React from 'react';

import baseStyle from '../baseStyle';
import computeDataAttributes from '../computeDataAttributes';
import validateChildrenProp from '../validateChildrenProp';
import Small from './Small';
import Strong from './Strong';

export interface TextProps {
  children: string | (string | React.ReactElement)[];
  className?: string;
  dataAttributes?: { [key: string]: string };
}

const textStyle = css(baseStyle, {
  fontSize: tokens.size.font[300].value,
  fontWeight: tokens.size.fontWeight.regular.value,
  lineHeight: tokens.size.font.base.value,
});

const Text: React.FC<TextProps> = (props: TextProps) => {
  validateChildrenProp(props, 'Text', [Small, Strong]);
  const { children, className, dataAttributes } = props;
  const parsedDataAttributes = computeDataAttributes(dataAttributes);

  return (
    <span className={className} css={textStyle} {...parsedDataAttributes}>
      {children}
    </span>
  );
};

export default Text;
