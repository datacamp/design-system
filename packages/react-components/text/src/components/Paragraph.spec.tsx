import '@testing-library/jest-dom/extend-expect';

import axeRender from '@datacamp/waffles-axe-render';
import React from 'react';

import Emphasis from './Emphasis';
import Link from './Link';
import Paragraph from './Paragraph';
import Small from './Small';
import Strong from './Strong';
import Text from './Text';

const testText = 'Some example text.';
const testClassName = 'example-class';
const testDataAttributes = { cy: 'othertest', example: 'testdata' };

describe('<Paragraph />', () => {
  it('renders with a string', async () => {
    const { container } = await axeRender(
      <Paragraph className={testClassName} dataAttributes={testDataAttributes}>
        {testText}
      </Paragraph>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders with Strong, Small, Text, Link and Emphasis children', async () => {
    const { container } = await axeRender(
      <Paragraph className={testClassName} dataAttributes={testDataAttributes}>
        Some example text <Strong>with strong</Strong> included, and also some{' '}
        <Text>custom text</Text>. It has some <Small>small text</Small> and{' '}
        <Emphasis>emphasis text</Emphasis>. It also contains{' '}
        <Link href="https://datacamp.com">a link</Link>.
      </Paragraph>
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('sets the text in the span', async () => {
    const { getByText } = await axeRender(<Paragraph>{testText}</Paragraph>);
    expect(getByText(testText)).toBeDefined();
  });

  it('sets the className', async () => {
    const { getByText } = await axeRender(
      <Paragraph className={testClassName}>{testText}</Paragraph>
    );
    expect(getByText(testText)).toHaveClass(testClassName);
  });

  it('sets the data attributes', async () => {
    const { getByText } = await axeRender(
      <Paragraph dataAttributes={testDataAttributes}>{testText}</Paragraph>
    );
    expect(getByText(testText)).toHaveAttribute(
      'data-cy',
      testDataAttributes.cy
    );
    expect(getByText(testText)).toHaveAttribute(
      'data-example',
      testDataAttributes.example
    );
  });
});
