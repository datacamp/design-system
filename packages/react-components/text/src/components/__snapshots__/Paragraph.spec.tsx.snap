// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Paragraph /> renders with Strong, Small, Text, Link and Emphasis children 1`] = `
.emotion-5 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-size: 16px;
  font-weight: 400;
  line-height: 1.5;
  margin: 0;
}

*:not(style) ~ .emotion-5 {
  margin-top: 8px;
}

.emotion-0 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-weight: 700;
  line-height: 16px;
}

.emotion-1 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-size: 16px;
  font-weight: 400;
  line-height: 16px;
}

.emotion-2 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-size: 14px;
  line-height: 16px;
}

.emotion-3 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-style: italic;
  font-weight: 400;
  line-height: 16px;
}

.emotion-4 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  color: #33aacc;
  cursor: pointer;
  font-size: 16px;
  font-weight: 400;
  line-height: 16px;
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<p
  class="example-class emotion-5"
  data-example="testdata"
  data-item="othertest"
>
  Some example text 
  <strong
    class="emotion-0"
  >
    with strong
  </strong>
   included, and also some
   
  <span
    class="emotion-1"
  >
    custom text
  </span>
  . It has some 
  <small
    class="emotion-2"
  >
    small text
  </small>
   and
   
  <em
    class="emotion-3"
  >
    emphasis text
  </em>
  . It also contains
   
  <a
    class="emotion-4"
    href="https://datacamp.com"
  >
    a link
  </a>
  .
</p>
`;

exports[`<Paragraph /> renders with a string 1`] = `
.emotion-0 {
  -webkit-font-smoothing: antialiased;
  color: rgb(77,83,86);
  font-family: sans-serif;
  font-family: Lato;
  font-style: normal;
  font-size: 16px;
  font-weight: 400;
  line-height: 1.5;
  margin: 0;
}

*:not(style) ~ .emotion-0 {
  margin-top: 8px;
}

<p
  class="example-class emotion-0"
  data-example="testdata"
  data-item="othertest"
>
  Some example text.
</p>
`;
